packages/compile-common/src/compilerSupplier/constructor.ts

```diff
@@ -0,0 +1,35 @@
+export namespace Constructor {
+  /**
+   * Type-level description for a particular constructor
+   */
+  export type Specification = {
+    /**
+     * Options argument for specified constructor
+     */
+    options: any;
+
+    /**
+     * Output object created by specified constructor
+     */
+    result: any;
+  };
+
+  /**
+   * Getter type for constructor arg
+   */
+  export type Options<S extends Specification> = S["options"];
```

Comment:
So basically you're making it look like there's something nice and generic going on here when actually it's all manual. :P

Reply:
sure... we'll go with that?